### ISSUE TITLE
Provide channel conversions

### ISSUE THREAD

**Issue by tindzk** [MetaStack-pl/MetaRx#1]
We should provide conversions from ``Future``, ``Option`` and other monadic types to ``Channel``.

**Comment by mkotsbak** [comment]
It would allow something similar to this in Scala.Rx (where Model.findXXX() returns Future[U]):

val results: Rx[List[U]] = Rx { Model.findXXX(query()).async(List.empty[U], discardLate = true)

**Comment by tindzk** [comment]
We could add the following three implicit conversions for ``Future``:

```scala
Future[Seq[T]] -> Buffer[T]
Future[Option[T]] -> Opt[T]
Future[T] -> Var[T]
```

Given a function ``Model.findX: Future[Seq[String]]``, you could write:

```scala
val results: ReadBuffer[String] = query.flatMapBuf(Model.findX)
```

Do you see a better way to implement this?

**Comment by mkotsbak** [comment]
That would been a nice solution yes!

**Comment by mkotsbak** [comment]
I think I like the monadic way better than all Rx {} wrapping in Scala.Rx.

**Comment by mkotsbak** [comment]
Maybe also a constructor like Buffer( Channel[Future[Seq[T]], discardLate: Boolean) ?

**Comment by tindzk** [comment]
Are you sure about ``Channel[Future[Seq[T]]]``? What is ``discardLate`` for?

**Comment by mkotsbak** [comment]
See https://github.com/lihaoyi/scala.rx/blob/master/scalarx/shared/src/main/scala/rx/ops/Async.scala
"The Async can be configured with a variety of Targets, to configure its handling of Futures which complete out of order (RunAlways, DiscardLate)".

**Comment by mkotsbak** [comment]
Maybe Future->Buffer conversion should be possible in different ways also, depending if you have handled errors yourself (using Future.recover..) or want to handle it yourself (generate Buffer[Try[SomeType]]),

something like

Buffer.fromFutureSuccess(Future[T]) : Buffer[T]
Buffer.fromFutureWithErrors(Future[T]): Buffer[Try[T]]

(or make the first just the default conversion and call it Buffer.fromFuture).

**Comment by tindzk** [comment]
I'm a bit wary about adding a lot of implicit conversions. That's why I haven't tackled this issue yet.

After thinking more about this issue, I believe that ``Future``s can be seen as one of Scala's core features. As we provide default constructors for ``Seq``, it would make sense to add similar ones for ``Future``s too. For example:

```scala
Buffer(future: Future[Seq[T]])
Opt(future: Future[Option[T]])
Var(future: Future[T])
```

I don't want to add special handling for error cases as you suggested since ``Future[Seq[T]]`` has only one error state, so `Buffer[Try[T]]` would imply we'd end up with one element in the error case. In the success case, we'd have to wrap up every element in ``Try[T]``. Error handling can still be done separately by calling `onFailure` on the ``Future``.

**Comment by mkotsbak** [comment]
Yes, I agree. It should be possible to handle in place before converting it to Buffer.

**Comment by mkotsbak** [comment]
ReadChannel[Option[T]] should also be converted to Opt[T]?

**Comment by tindzk** [comment]
You can call `state` on `Channel` (or `cache` on `ReadChannel`). `Opt` should only be returned if you expect that a value will be produced on the channel. As for `ReadChannel` there's nothing you could do with those produced values. Back-propagation would only work on `Channel`.

**Comment by mkotsbak** [comment]
Hmm, could a ReadChannel[Option[T]] be converted to a ReadPartialChannel[T] then? It would be more convenient to handle.

**Comment by tindzk** [comment]
Good idea. I implemented it in 9b5af8fde1276fca266298ff0f9d3de1547401d9. If you have any idea how to solve it better, please let me know.

**Comment by mkotsbak** [comment]
Nice!
